--- ../bacula-web/classes.inc	2006-03-14 03:36:00.000000000 -0500
+++ ../bacula-web/classes.inc	2006-03-14 03:30:14.000000000 -0500
@@ -155,7 +155,7 @@
                                 return 0;
                 }
                 else if ( $this->driver == "pgsql") {
-                        $dbsize = $this->link->query("select pg_database_size('$this->dbs_name')")
+                        $dbsize = $this->link->query("select database_size('$this->dbs_name')")
                                 or die ("classes.inc: Error query: 4");
 			if (PEAR::isError($dbsize))
 	                    die($dbsize->getMessage());
