--- linux-2.6.20.noarch/arch/x86_64/kernel/time-xen.c~orig	2007-04-26 02:05:31.000000000 -0700
+++ linux-2.6.20.noarch/arch/x86_64/kernel/time-xen.c	2007-04-26 02:02:27.000000000 -0700
@@ -985,7 +985,7 @@ void time_resume(void)
 #ifdef CONFIG_SMP
 static char timer_name[NR_CPUS][15];
 
-void local_setup_timer(unsigned int cpu)
+int local_setup_timer(unsigned int cpu)
 {
 	int seq;
 
@@ -1009,6 +1009,7 @@ void local_setup_timer(unsigned int cpu)
 			timer_name[cpu],
 			NULL);
 	BUG_ON(per_cpu(timer_irq, cpu) < 0);
+	return 0;
 }
 
 void local_teardown_timer(unsigned int cpu)
