diff --git a/arch/i386/kernel/head-xen.S b/arch/i386/kernel/head-xen.S
index 07d2dc2..f4237fd 100644
--- a/arch/i386/kernel/head-xen.S
+++ b/arch/i386/kernel/head-xen.S
@@ -248,7 +248,7 @@ ENTRY(cpu_gdt_table)
 	.ascii	",ELF_PADDR_OFFSET=0x"
 		utoa __PAGE_OFFSET
 	.ascii	",VIRT_ENTRY=0x"
-		utoa (__PAGE_OFFSET + LOAD_PHYSICAL_ADDR + VIRT_ENTRY_OFFSET)
+		utoa (__PAGE_OFFSET+LOAD_PHYSICAL_ADDR+VIRT_ENTRY_OFFSET)
 	.ascii	",HYPERCALL_PAGE=0x"
 		utoa ((LOAD_PHYSICAL_ADDR+HYPERCALL_PAGE_OFFSET)>>PAGE_SHIFT)
 	.ascii  ",FEATURES=writable_page_tables"
diff --git a/arch/x86_64/kernel/head-xen.S b/arch/x86_64/kernel/head-xen.S
index e847d47..2d34c5d 100644
--- a/arch/x86_64/kernel/head-xen.S
+++ b/arch/x86_64/kernel/head-xen.S
@@ -163,9 +163,9 @@ ENTRY(empty_zero_page)
 	.ascii	",ELF_PADDR_OFFSET=0x"
 		utoh __START_KERNEL_map
 	.ascii	",VIRT_ENTRY=0x"
-		utoh (__START_KERNEL_map + __PHYSICAL_START + VIRT_ENTRY_OFFSET)
+		utoh (__START_KERNEL_map+__PHYSICAL_START+VIRT_ENTRY_OFFSET)
 	.ascii	",HYPERCALL_PAGE=0x"
-		utoh (phys_hypercall_page >> PAGE_SHIFT)
+		utoh (phys_hypercall_page>>PAGE_SHIFT)
 	.ascii  ",FEATURES=writable_page_tables"
 	.ascii		 "|writable_descriptor_tables"
 	.ascii		 "|auto_translated_physmap"
diff --git a/include/asm-i386/boot.h b/include/asm-i386/boot.h
index e7686d0..8fae4d6 100644
--- a/include/asm-i386/boot.h
+++ b/include/asm-i386/boot.h
@@ -13,8 +13,6 @@
 #define ASK_VGA		0xfffd		/* ask for it at bootup */
 
 /* Physical address where kenrel should be loaded. */
-#define LOAD_PHYSICAL_ADDR ((CONFIG_PHYSICAL_START \
-				+ (CONFIG_PHYSICAL_ALIGN - 1)) \
-				& ~(CONFIG_PHYSICAL_ALIGN - 1))
+#define LOAD_PHYSICAL_ADDR ((CONFIG_PHYSICAL_START+(CONFIG_PHYSICAL_ALIGN-1))&~(CONFIG_PHYSICAL_ALIGN-1))
 
 #endif /* _LINUX_BOOT_H */
