diff -urN bacula-2.2.4.orig/scripts/bacula.desktop.gnome1.xsu.in bacula-2.2.4/scripts/bacula.desktop.gnome1.xsu.in
--- bacula-2.2.4.orig/scripts/bacula.desktop.gnome1.xsu.in	2007-04-18 19:06:43.000000000 +0200
+++ bacula-2.2.4/scripts/bacula.desktop.gnome1.xsu.in	2007-10-10 23:34:42.000000000 +0200
@@ -2,7 +2,7 @@
 Name=Bacula Console
 Comment=Bacula Director Console
 Icon=/usr/share/pixmaps/bacula.png
-Exec=gnomesu -t "Query" -c "@sbindir@/bgnome-console -c @sysconfdir@/bgnome-console.conf" -d -e -m "In order to run the bacula console as root, ^additional information is required."
+Exec=gksu -u root -m 'Bacula Console needs to be run with superuser privileges' "@sbindir@/bgnome-console -c @sysconfdir@/bgnome-console.conf"
 Terminal=false
 Type=Application
 Encoding=UTF-8
diff -urN bacula-2.2.4.orig/scripts/bacula.desktop.gnome2.xsu.in bacula-2.2.4/scripts/bacula.desktop.gnome2.xsu.in
--- bacula-2.2.4.orig/scripts/bacula.desktop.gnome2.xsu.in	2007-04-18 19:06:43.000000000 +0200
+++ bacula-2.2.4/scripts/bacula.desktop.gnome2.xsu.in	2007-10-10 23:32:44.000000000 +0200
@@ -2,7 +2,7 @@
 Name=Bacula Console
 Comment=Bacula Director Console
 Icon=/usr/share/pixmaps/bacula.png
-Exec=gnomesu -t "Query" -c "@sbindir@/bgnome-console -c @sysconfdir@/bgnome-console.conf" -d -e -m "In order to run the bacula console as root, ^additional information is required."
+Exec=gksu -u root -m 'Bacula Console needs to be run with superuser privileges' "@sbindir@/bgnome-console -c @sysconfdir@/bgnome-console.conf"
 Terminal=false
 Type=Application
 Encoding=UTF-8
diff -urN bacula-2.2.4.orig/scripts/bat.desktop.xsu.in bacula-2.2.4/scripts/bat.desktop.xsu.in
--- bacula-2.2.4.orig/scripts/bat.desktop.xsu.in	2007-04-29 15:48:18.000000000 +0200
+++ bacula-2.2.4/scripts/bat.desktop.xsu.in	2007-10-10 23:33:57.000000000 +0200
@@ -2,7 +2,7 @@
 Name=Bacula Administration Tool
 Comment=Bacula Director Console
 Icon=/usr/share/pixmaps/bat_icon.png
-Exec=gnomesu -t "Query" -c "@sbindir@/bat -c @sysconfdir@/bat.conf" -d -e -m "In order to run the bacula console as root, ^additional information is required."
+Exec=gksu -u root -m 'Bacula Administration Tool needs to be run with superuser privileges' "@sbindir@/bat -c @sysconfdir@/bat.conf"
 Terminal=false
 Type=Application
 Encoding=UTF-8
diff -urN bacula-2.2.4.orig/scripts/wxconsole.desktop.xsu.in bacula-2.2.4/scripts/wxconsole.desktop.xsu.in
--- bacula-2.2.4.orig/scripts/wxconsole.desktop.xsu.in	2007-04-29 15:11:37.000000000 +0200
+++ bacula-2.2.4/scripts/wxconsole.desktop.xsu.in	2007-10-10 23:34:06.000000000 +0200
@@ -2,7 +2,7 @@
 Name=Bacula WX Console
 Comment=Bacula Director Console
 Icon=/usr/share/pixmaps/wxwin16x16.xpm
-Exec=gnomesu -t "Query" -c "@sbindir@/bwx-console -c @sysconfdir@/bwx-console.conf" -d -e -m "In order to run the bacula wx console as root, ^additional information is required."
+Exec=gksu -u root -m 'Bacula WX Console needs to be run with superuser privileges' "@sbindir@/bwx-console -c @sysconfdir@/bwx-console.conf"
 Terminal=false
 Type=Application
 Encoding=UTF-8
