--- src/lxc/Makefile.am.old	2013-06-08 13:32:02.927550207 +0200
+++ src/lxc/Makefile.am	2013-06-08 13:32:59.020579658 +0200
@@ -30,11 +30,9 @@
 endif
 endif
 
-sodir=$(libdir)
-# use PROGRAMS to avoid complains from automake
-so_PROGRAMS = liblxc.so
+lib_LTLIBRARIES = liblxc-@VERSION@.la
 
-liblxc_so_SOURCES = \
+liblxc_@VERSION@_la_SOURCES = \
 	arguments.c arguments.h \
 	commands.c commands.h \
 	start.c start.h \
@@ -76,14 +74,14 @@
 	version.c version.h
 
 if IS_BIONIC
-liblxc_so_SOURCES += \
+liblxc_@VERSION@_la_SOURCES += \
 	../include/openpty.c ../include/openpty.h \
 	../include/lxcmntent.c ../include/lxcmntent.h
 endif
 
 if !HAVE_GETLINE
 if HAVE_FGETLN
-liblxc_so_SOURCES += ../include/getline.c ../include/getline.h
+liblxc_@VERSION@_la_SOURCES += ../include/getline.c ../include/getline.h
 endif
 endif
 
@@ -105,16 +103,13 @@
 
 if ENABLE_SECCOMP
 AM_CFLAGS += -DHAVE_SECCOMP
-liblxc_so_SOURCES += seccomp.c
+liblxc_@VERSION@_la_SOURCES += seccomp.c
 endif
 
-liblxc_so_CFLAGS = -fPIC -DPIC $(AM_CFLAGS)
+liblxc_@VERSION@_la_LDFLAGS = \
+	-avoid-version
 
-liblxc_so_LDFLAGS = \
-	-shared \
-	-Wl,-soname,liblxc.so.$(firstword $(subst ., ,$(VERSION)))
-
-liblxc_so_LDADD = $(CAP_LIBS) $(APPARMOR_LIBS) $(SECCOMP_LIBS)
+liblxc_@VERSION@_la_LIBADD = -lutil $(CAP_LIBS) $(APPARMOR_LIBS) $(SECCOMP_LIBS)
 
 bin_SCRIPTS = \
 	lxc-ps \
@@ -166,11 +161,7 @@
 #pkglibexec_SCRIPTS = \
 #	lxc.functions
 
-AM_LDFLAGS = -Wl,-E
-if ENABLE_RPATH
-AM_LDFLAGS += -Wl,-rpath -Wl,$(libdir)
-endif
-LDADD=liblxc.so @CAP_LIBS@ @APPARMOR_LIBS@ @SECCOMP_LIBS@
+LDADD=liblxc-@VERSION@.la
 
 lxc_attach_SOURCES = lxc_attach.c
 lxc_cgroup_SOURCES = lxc_cgroup.c
@@ -180,6 +171,7 @@
 lxc_freeze_SOURCES = lxc_freeze.c
 lxc_info_SOURCES = lxc_info.c
 lxc_init_SOURCES = lxc_init.c
+lxc_init_LDFLAGS = -all-static
 lxc_monitor_SOURCES = lxc_monitor.c
 lxc_restart_SOURCES = lxc_restart.c
 lxc_start_SOURCES = lxc_start.c
@@ -189,14 +181,5 @@
 lxc_wait_SOURCES = lxc_wait.c
 lxc_kill_SOURCES = lxc_kill.c
 
-install-exec-local: install-soPROGRAMS
-	mkdir -p $(DESTDIR)$(datadir)/lxc
+install-exec-local:
 	install -c -m 644 lxc.functions $(DESTDIR)$(datadir)/lxc
-	mv $(DESTDIR)$(libdir)/liblxc.so $(DESTDIR)$(libdir)/liblxc.so.$(VERSION)
-	/sbin/ldconfig -l $(DESTDIR)$(libdir)/liblxc.so.$(VERSION)
-	cd $(DESTDIR)$(libdir); \
-	ln -sf liblxc.so.$(VERSION) liblxc.so.$(firstword $(subst ., ,$(VERSION))); \
-	ln -sf liblxc.so.$(firstword $(subst ., ,$(VERSION))) liblxc.so
-
-uninstall-local:
-	$(RM) $(DESTDIR)$(libdir)/liblxc.so*
--- configure.ac.old	2013-06-08 13:29:41.148318519 +0200
+++ configure.ac	2013-06-08 13:30:53.461259871 +0200
@@ -52,11 +52,7 @@
 AC_MSG_RESULT([$with_distro])
 AM_CONDITIONAL([HAVE_DEBIAN], [test x"$with_distro" = "xdebian" -o x"$with_distro" = "xubuntu"])
 
-# Allow disabling rpath
-AC_ARG_ENABLE([rpath],
-	[AC_HELP_STRING([--disable-rpath], [do not set rpath in executables])],
-	[], [enable_rpath=yes])
-AM_CONDITIONAL([ENABLE_RPATH], [test "x$enable_rpath" = "xyes"])
+LT_INIT
 
 # Documentation (manpages)
 AC_ARG_ENABLE([doc],
