diff --git a/fs/dcache.c b/fs/dcache.c
index b69a57e..f32cef2 100644
--- a/fs/dcache.c
+++ b/fs/dcache.c
@@ -231,14 +231,14 @@ kill_it:
 	ub = dentry->dentry_bc.d_ub;
 	d_ubsize = dentry->dentry_bc.d_ubsize;
 	dentry = d_kill(dentry);
+	preempt_disable();
 	if (ub_dentry_on) {
 		uncharge_dcache(ub, d_ubsize);
 		put_beancounter(ub);
 	}
-	if (dentry) {
-		preempt_disable();
+	if (dentry) 
 		goto repeat;
-	}
+	preempt_enable();
 }
 
 void dput(struct dentry *dentry)
