--- mytharchive/mythburn/scripts/mythburn.py	2006-12-23 10:45:26.000000000 -0500
+++ /usr/share/mythtv/mytharchive/scripts/mythburn.py	2006-12-23 12:24:09.000000000 -0500
@@ -25,6 +25,7 @@
 
 #Optional (only needed for tcrequant)
 #transcode - v1.0.2
+#projectx - 0.90.3.00
 
 #******************************************************************************
 #******************************************************************************
@@ -3907,6 +3908,7 @@
 path_jpeg2yuv = [defaultsettings["MythArchiveJpeg2yuvCmd"], os.path.split(defaultsettings["MythArchiveJpeg2yuvCmd"])[1]]
 path_spumux = [defaultsettings["MythArchiveSpumuxCmd"], os.path.split(defaultsettings["MythArchiveSpumuxCmd"])[1]]
 path_mpeg2enc = [defaultsettings["MythArchiveMpeg2encCmd"], os.path.split(defaultsettings["MythArchiveMpeg2encCmd"])[1]]
+path_projectx = [defaultsettings["MythArchiveProjectXCmd"], os.path.split(defaultsettings["MythArchiveProjectXCmd"])[1]]
 
 try:
     try:
