--- console/console.c	2007-08-11 23:50:20.000000000 +0200
+++ console/console.c	2007-06-09 11:25:27.000000000 +0200
@@ -105,7 +105,7 @@
 static int execcmd(FILE *input, BSOCK *UA_sock);
 
 
-#define CONFIG_FILE "bconsole.conf"   /* default configuration file */
+#define CONFIG_FILE "/etc/bacula/bconsole.conf"   /* default configuration file */
 
 static void usage()
 {
--- dird/dird.c	2007-08-11 23:49:40.000000000 +0200
+++ dird/dird.c	2007-06-07 16:46:43.000000000 +0200
@@ -85,7 +85,7 @@
 extern URES res_all;
 #endif
 
-#define CONFIG_FILE "bacula-dir.conf" /* default configuration file */
+#define CONFIG_FILE "/etc/bacula/bacula-dir.conf" /* default configuration file */
 
 static void usage()
 {
--- filed/filed.c	2007-08-11 23:51:18.000000000 +0200
+++ filed/filed.c	2007-06-07 16:46:43.000000000 +0200
@@ -50,7 +50,7 @@
 void *start_heap;
 
 
-#define CONFIG_FILE "bacula-fd.conf" /* default config file */
+#define CONFIG_FILE "/etc/bacula/bacula-fd.conf" /* default config file */
 
 char *configfile = NULL;
 static bool foreground = false;
--- gnome2-console/console.c	2007-08-11 23:50:45.000000000 +0200
+++ gnome2-console/console.c	2007-06-07 16:46:43.000000000 +0200
@@ -93,7 +93,7 @@
 static guint initial;
 static int numdir = 0;
 
-#define CONFIG_FILE "./bgnome-console.conf"   /* default configuration file */
+#define CONFIG_FILE "/etc/bacula/bgnome-console.conf"   /* default configuration file */
 
 static void usage()
 {
--- qt-console/main.cpp	2007-08-11 23:55:10.000000000 +0200
+++ qt-console/main.cpp	2007-06-07 16:46:43.000000000 +0200
@@ -46,7 +46,7 @@
 static void usage();
 static int check_resources();
 
-#define CONFIG_FILE "./bat.conf"   /* default configuration file */
+#define CONFIG_FILE "/etc/bacula/bat.conf"   /* default configuration file */
 
 /* Static variables */
 static char *configfile = NULL;
--- stored/bcopy.c	2007-08-11 23:54:48.000000000 +0200
+++ stored/bcopy.c	2007-06-07 16:46:43.000000000 +0200
@@ -57,7 +57,7 @@
 static uint32_t jobs = 0;
 static DEV_BLOCK *out_block;
 
-#define CONFIG_FILE "bacula-sd.conf"
+#define CONFIG_FILE "/etc/bacula/bacula-sd.conf"
 char *configfile = NULL;
 STORES *me = NULL;                    /* our Global resource */
 bool forge_on = false;                /* proceed inspite of I/O errors */
--- stored/bextract.c	2007-08-11 23:45:44.000000000 +0200
+++ stored/bextract.c	2007-06-07 16:46:43.000000000 +0200
@@ -64,7 +64,7 @@
 static uint32_t wsize;                /* write size */
 static uint64_t fileAddr = 0;         /* file write address */
 
-#define CONFIG_FILE "bacula-sd.conf"
+#define CONFIG_FILE "/etc/bacula/bacula-sd.conf"
 char *configfile = NULL;
 STORES *me = NULL;                    /* our Global resource */
 bool forge_on = false;
--- stored/bls.c	2007-08-11 23:46:05.000000000 +0200
+++ stored/bls.c	2007-06-07 16:46:43.000000000 +0200
@@ -59,7 +59,7 @@
 static uint32_t num_files = 0;
 static ATTR *attr;
 
-#define CONFIG_FILE "bacula-sd.conf"
+#define CONFIG_FILE "/etc/bacula/bacula-sd.conf"
 char *configfile = NULL;
 STORES *me = NULL;                    /* our Global resource */
 bool forge_on = false;
--- stored/bscan.c	2007-08-11 23:46:26.000000000 +0200
+++ stored/bscan.c	2007-06-18 21:29:26.000000000 +0200
@@ -100,7 +100,7 @@
 static int num_media = 0;
 static int num_files = 0;
 
-#define CONFIG_FILE "bacula-sd.conf"
+#define CONFIG_FILE "/etc/bacula/bacula-sd.conf"
 char *configfile = NULL;
 STORES *me = NULL;                    /* our Global resource */
 bool forge_on = false;                /* proceed inspite of I/O errors */
--- stored/btape.c	2007-08-11 23:46:59.000000000 +0200
+++ stored/btape.c	2007-06-07 16:46:43.000000000 +0200
@@ -98,7 +98,7 @@
 
 
 /* Static variables */
-#define CONFIG_FILE "bacula-sd.conf"
+#define CONFIG_FILE "/etc/bacula/bacula-sd.conf"
 char *configfile = NULL;
 
 #define MAX_CMD_ARGS 30
--- stored/stored.c	2007-08-11 23:54:32.000000000 +0200
+++ stored/stored.c	2007-06-23 11:58:34.000000000 +0200
@@ -52,7 +52,7 @@
 
 extern "C" void *device_initialization(void *arg);
 
-#define CONFIG_FILE "bacula-sd.conf"  /* Default config file */
+#define CONFIG_FILE "/etc/bacula/bacula-sd.conf"  /* Default config file */
 
 /* Global variables exported */
 char OK_msg[]   = "3000 OK\n";
--- tray-monitor/tray-monitor.c	2007-08-11 23:49:13.000000000 +0200
+++ tray-monitor/tray-monitor.c	2007-06-07 16:46:43.000000000 +0200
@@ -104,7 +104,7 @@
 
 PangoFontDescription *font_desc = NULL;
 
-#define CONFIG_FILE "./tray-monitor.conf"   /* default configuration file */
+#define CONFIG_FILE "/etc/bacula/tray-monitor.conf"   /* default configuration file */
 
 static void usage()
 {
