--- stored/bcopy.c	2006-11-27 05:03:01.000000000 -0500
+++ stored/bcopy.c	2007-01-11 21:14:58.883253250 -0500
@@ -57,7 +57,7 @@
 static uint32_t jobs = 0;
 static DEV_BLOCK *out_block;
 
-#define CONFIG_FILE "bacula-sd.conf"
+#define CONFIG_FILE "/etc/bacula/bacula-sd.conf"
 char *configfile = NULL;
 STORES *me = NULL;                    /* our Global resource */
 bool forge_on = false;                /* proceed inspite of I/O errors */
--- stored/bextract.c	2006-12-01 03:45:13.000000000 -0500
+++ stored/bextract.c	2007-01-11 21:13:33.333906750 -0500
@@ -64,7 +64,7 @@
 static uint32_t wsize;                /* write size */
 static uint64_t fileAddr = 0;         /* file write address */
 
-#define CONFIG_FILE "bacula-sd.conf"
+#define CONFIG_FILE "/etc/bacula/bacula-sd.conf"
 char *configfile = NULL;
 STORES *me = NULL;                    /* our Global resource */
 bool forge_on = false;
--- stored/bls.c	2006-12-13 14:42:12.000000000 -0500
+++ stored/bls.c	2007-01-11 21:13:47.850814000 -0500
@@ -59,7 +59,7 @@
 static uint32_t num_files = 0;
 static ATTR *attr;
 
-#define CONFIG_FILE "bacula-sd.conf"
+#define CONFIG_FILE "/etc/bacula/bacula-sd.conf"
 char *configfile = NULL;
 STORES *me = NULL;                    /* our Global resource */
 bool forge_on = false;
--- stored/bscan.c	2006-12-01 03:45:13.000000000 -0500
+++ stored/bscan.c	2007-01-11 21:14:05.375909250 -0500
@@ -100,7 +100,7 @@
 static int num_media = 0;
 static int num_files = 0;
 
-#define CONFIG_FILE "bacula-sd.conf"
+#define CONFIG_FILE "/etc/bacula/bacula-sd.conf"
 char *configfile = NULL;
 STORES *me = NULL;                    /* our Global resource */
 bool forge_on = false;                /* proceed inspite of I/O errors */
--- stored/btape.c	2006-12-14 06:41:00.000000000 -0500
+++ stored/btape.c	2007-01-11 21:14:29.265402250 -0500
@@ -98,7 +98,7 @@
 
 
 /* Static variables */
-#define CONFIG_FILE "bacula-sd.conf"
+#define CONFIG_FILE "/etc/bacula/bacula-sd.conf"
 char *configfile = NULL;
 
 #define MAX_CMD_ARGS 30
--- tray-monitor/tray-monitor.c	2006-12-01 03:45:14.000000000 -0500
+++ tray-monitor/tray-monitor.c	2007-01-11 21:17:29.828686750 -0500
@@ -101,7 +101,7 @@
 char** xpm_generic_var;
 static gboolean blinkstate = TRUE;
 
-#define CONFIG_FILE "./tray-monitor.conf"   /* default configuration file */
+#define CONFIG_FILE "/etc/bacula/tray-monitor.conf"   /* default configuration file */
 
 static void usage()
 {
--- dird/dird.c	2006-12-22 10:01:05.000000000 -0500
+++ dird/dird.c	2007-01-11 21:18:16.255588250 -0500
@@ -83,7 +83,7 @@
 extern URES res_all;
 #endif
 
-#define CONFIG_FILE "bacula-dir.conf" /* default configuration file */
+#define CONFIG_FILE "/etc/bacula/bacula-dir.conf" /* default configuration file */
 
 static void usage()
 {
--- console/console.c	2006-11-27 05:02:59.000000000 -0500
+++ console/console.c	2007-01-11 21:18:36.580858500 -0500
@@ -104,7 +104,7 @@
 static int sleepcmd(FILE *input, BSOCK *UA_sock);
 
 
-#define CONFIG_FILE "bconsole.conf"   /* default configuration file */
+#define CONFIG_FILE "/etc/bacula/bconsole.conf"   /* default configuration file */
 
 static void usage()
 {
--- gnome2-console/console.c	2006-12-08 09:27:10.000000000 -0500
+++ gnome2-console/console.c	2007-01-11 21:19:17.287402500 -0500
@@ -93,7 +93,7 @@
 static guint initial;
 static int numdir = 0;
 
-#define CONFIG_FILE "./gnome-console.conf"   /* default configuration file */
+#define CONFIG_FILE "/etc/bacula/gnome-console.conf"   /* default configuration file */
 
 static void usage()
 {
--- filed/filed.c	2006-11-27 05:03:00.000000000 -0500
+++ filed/filed.c	2007-01-11 21:20:13.722929500 -0500
@@ -49,7 +49,7 @@
 bool no_signals = false;
 
 
-#define CONFIG_FILE "bacula-fd.conf" /* default config file */
+#define CONFIG_FILE "/etc/bacula/bacula-fd.conf" /* default config file */
 
 char *configfile = NULL;
 static bool foreground = false;
