diff -urN scripts/bacula.desktop.gnome1.xsu.in scripts/bacula.desktop.gnome1.xsu.in
--- scripts/bacula.desktop.gnome1.xsu.in	2004-03-28 19:08:45.000000000 +0200
+++ scripts/bacula.desktop.gnome1.xsu.in	2007-02-28 16:36:14.000000000 +0100
@@ -2,7 +2,7 @@
 Name=Bacula Console
 Comment=Bacula Director Console
 Icon=/usr/share/pixmaps/bacula.png
-Exec=gnomesu -t "Query" -c "@sbindir@/gnome-console -c @sysconfdir@/gnome-console.conf" -d -e -m "In order to run the bacula console as root, ^additional information is required."
+Exec=gksu -u root -m 'Bacula Console needs to be run with superuser privileges' "@sbindir@/gnome-console -c @sysconfdir@/gnome-console.conf"
 Terminal=false
 Type=Application
 Encoding=UTF-8
diff -urN scripts/bacula.desktop.gnome2.xsu.in scripts/bacula.desktop.gnome2.xsu.in
--- scripts/bacula.desktop.gnome2.xsu.in	2004-03-28 19:08:45.000000000 +0200
+++ scripts/bacula.desktop.gnome2.xsu.in	2007-02-28 16:36:25.000000000 +0100
@@ -2,7 +2,7 @@
 Name=Bacula Console
 Comment=Bacula Director Console
 Icon=/usr/share/pixmaps/bacula.png
-Exec=gnomesu -t "Query" -c "@sbindir@/gnome-console -c @sysconfdir@/gnome-console.conf" -d -e -m "In order to run the bacula console as root, ^additional information is required."
+Exec=gksu -u root -m 'Bacula Console needs to be run with superuser privileges' "@sbindir@/gnome-console -c @sysconfdir@/gnome-console.conf"
 Terminal=false
 Type=Application
 Encoding=UTF-8
diff -urN scripts/wxconsole.desktop.xsu.in scripts/wxconsole.desktop.xsu.in
--- scripts/wxconsole.desktop.xsu.in	2006-04-16 18:01:52.000000000 +0200
+++ scripts/wxconsole.desktop.xsu.in	2007-02-28 16:35:25.000000000 +0100
@@ -2,7 +2,7 @@
 Name=Bacula WX Console
 Comment=Bacula Director Console
 Icon=/usr/share/pixmaps/wxwin16x16.xpm
-Exec=gnomesu -t "Query" -c "@sbindir@/wxconsole -c @sysconfdir@/wxconsole.conf" -d -e -m "In order to run the bacula wx console as root, ^additional information is required."
+Exec=gksu -u root -m 'Bacula Console needs to be run with superuser privileges' "@sbindir@/wxconsole -c @sysconfdir@/wxconsole.conf"
 Terminal=false
 Type=Application
 Encoding=UTF-8
