--- mythweb/modules/mythburn/readlog.php	2006-03-08 16:37:40.000000000 -0500
+++ mythweb/modules/mythburn/readlog.php	2006-03-22 22:57:20.000000000 -0500
@@ -14,7 +14,7 @@
   
   //new 
   $logLinesArray = file($logfile);
-  $linesToDisplay = 15;
+  $linesToDisplay = 50;
   $loglinesCount = count($logLinesArray);
   //echo "logfile has $loglinesCount lines";
   $startindex = max(0,$loglinesCount-$linesToDisplay);
