--- partimage-0.6.5.orig/Makefile.in	2007-03-11 12:50:56.000000000 +0100
+++ partimage-0.6.5/Makefile.in	2007-03-11 12:56:16.000000000 +0100
@@ -669,8 +669,6 @@
 	rm ${sysconfdir}/partimaged/partimaged.csr ;\
 	chmod 600 ${sysconfdir}/partimaged/partimaged.key ;\
 	chmod 600 ${sysconfdir}/partimaged/partimaged.cert ;\
-	chown partimag:root ${sysconfdir}/partimaged/partimaged.key ;\
-	chown partimag:root ${sysconfdir}/partimaged/partimaged.cert ;\
 	else \
 	echo "SSL disabled, no certificate will be generated." ;\
 	fi
@@ -702,7 +700,6 @@
 	echo "#sample # user 'sample' is allowed to connect partimaged" >> \
            ${sysconfdir}/partimaged/partimagedusers ;\
 	chmod 600 ${sysconfdir}/partimaged/partimagedusers ;\
-	chown partimag:root ${sysconfdir}/partimaged/partimagedusers ;\
 	fi 
 
 	$(mkinstalldirs) ${infodir}
