--- /usr/portage/sys-block/partimage/partimage-0.6.4-r4.ebuild	2006-10-15 01:06:23.000000000 +0200
+++ partimage-0.6.5.ebuild	2007-03-11 14:05:43.000000000 +0100
@@ -1,27 +1,26 @@
-# Copyright 1999-2006 Gentoo Foundation
+# Copyright 1999-2007 Gentoo Foundation
 # Distributed under the terms of the GNU General Public License v2
-# $Header: /var/cvsroot/gentoo-x86/sys-block/partimage/partimage-0.6.4-r4.ebuild,v 1.14 2006/10/14 22:45:43 xmerlin Exp $
+# $Header: $
 
-WANT_AUTOMAKE="1.8"
+inherit eutils flag-o-matic pam
 
-inherit eutils flag-o-matic pam autotools
-
-MY_P="${PN}-${PV}-1"
 DESCRIPTION="Console-based application to efficiently save raw partition data to an image file. Optional encryption/compression support."
 HOMEPAGE="http://www.partimage.org/"
-SRC_URI="mirror://sourceforge/partimage/${MY_P}.tar.bz2"
+SRC_URI="mirror://sourceforge/partimage/${P}.tar.bz2"
 LICENSE="GPL-2"
 SLOT="0"
 KEYWORDS="~x86 ~ppc ~sparc"
-IUSE="ssl nologin nls pam static"
+IUSE="ssl nologin nls pam static unicode"
 
 DEPEND="virtual/libc
 	>=sys-libs/zlib-1.1.4
 	>=dev-libs/newt-0.51.6
 	app-arch/bzip2
+	sys-devel/automake
 	=sys-libs/slang-1*
 	nls? ( sys-devel/gettext )
-	ssl? ( >=dev-libs/openssl-0.9.6g )"
+	ssl? ( >=dev-libs/openssl-0.9.6g )
+	sys-devel/autoconf"
 
 RDEPEND="!static? ( virtual/libc
 		>=sys-libs/zlib-1.1.4
@@ -42,8 +41,6 @@
 PARTIMAG_USER_HOMEDIR=/var/log/partimage
 PARTIMAG_USER_GROUPS=partimag
 
-S="${WORKDIR}/${MY_P}"
-
 pkg_setup() {
 	# Now add users if needed
 	enewgroup ${PARTIMAG_GROUP_NAME} ${PARTIMAG_GROUP_GID}
@@ -55,22 +52,13 @@
 	cd ${S}
 
 	# we can do better security ourselves
-	epatch ${FILESDIR}/${P}-nodumbpermchecks.diff || die
-	epatch ${FILESDIR}/${P}-chown.patch || die
-	epatch ${FILESDIR}/${P}-not_install_info.patch || die
-	epatch ${FILESDIR}/${P}-fixserverargs.diff || die
-	epatch ${FILESDIR}/${P}-1-lib64.patch
-	epatch ${FILESDIR}/${P}-fflush-before-re-read-partition-table.patch || die
-	epatch ${FILESDIR}/${P}-save_all_and_rest_all_actions.patch || die
-	epatch ${FILESDIR}/${P}-datadir-path.patch || die
-	epatch ${FILESDIR}/${P}-dont-discard-error-message-in-batch-mode.patch || die
-	epatch ${FILESDIR}/${P}-save_file_and_rest_file_actions.patch || die
-	epatch ${FILESDIR}/${P}-varargs.patch || die
-	epatch ${FILESDIR}/${P}-gui.diff || die
-	epatch ${FILESDIR}/${P}-empty-salt.patch || die
-	epatch ${FILESDIR}/${P}-help.patch || die
-	epatch ${FILESDIR}/${P}-xfs.patch || die
-	epatch ${FILESDIR}/${P}-port.patch || die
+	epatch ${FILESDIR}/partimage-0.6.5-chown.patch || die
+	epatch ${FILESDIR}/partimage-0.6.4-not_install_info.patch || die
+	epatch ${FILESDIR}/partimage-0.6.4-datadir-path.patch || die
+
+	if use unicode; then
+		cp ${FILESDIR}/de.po ${S}/po/de.po
+	fi
 }
 
 src_compile() {
@@ -94,15 +82,15 @@
 		|| die "econf failed"
 
 	emake || die "make failed"
-	#if use pam
-	#then
-	#	make pamfile || die  "couldn't create pam file"
-	#fi
+	if use pam
+	then
+		make pamfile || die  "couldn't create pam file"
+	fi
 }
 
 src_install() {
 	emake DESTDIR=${D} \
-		MKINSTALLDIRS=/usr/share/automake-1.8/mkinstalldirs install || die
+		MKINSTALLDIRS=/usr/share/automake-1.10/mkinstalldirs install || die
 
 	keepdir /var/log/partimage
 
@@ -112,14 +100,11 @@
 	exeinto /etc/init.d ; newexe ${FILESDIR}/${PN}d.init ${PN}d || die
 	insinto /etc/conf.d ; newins ${FILESDIR}/${PN}d.conf ${PN}d || die
 
-	doman debian/partimage.1 debian/partimaged.8 ${FILESDIR}/partimagedusers.5 || die
+	doman doc/en/man/partimage.1 doc/en/man/partimaged.8 doc/en/man/partimagedusers.5 || die
 	dodoc AUTHORS BUGS COPYING ChangeLog INSTALL README* TODO partimage.lsm
 
 	# pam
-	if use pam
-	then
-		newpamd ${FILESDIR}/partimaged.pam partimaged || die
-	fi
+	newpamd partimaged.pam partimaged
 }
 
 # vars for SSL stuff
@@ -179,6 +164,6 @@
 		einfo "emerge  --config =${PF}"
 		# force a permmissions fixup
 		partimagesslperms
+		return 0
 	fi
-	chown partimag:0 /etc/partimaged/partimagedusers || die
 }
