--- src/xercesc/util/MsgLoaders/ICU/resources/Makefile.in.orig	2008-06-23 12:57:28.000000000 +0200
+++ src/xercesc/util/MsgLoaders/ICU/resources/Makefile.in	2009-05-22 20:32:59.000000000 +0200
@@ -50,8 +50,7 @@
 
 # Add -v if you want to see more verbose output.
 #
-PKGDATAOPTS=-d . -M '"CC=$(CC)" "CXX=$(CXX)" "AR=$(AR)" "RANLIB=$(RANLIB)" \
-"CPPFLAGS=$(CPPFLAGS)" "CFLAGS=$(CFLAGS)" "CXXFLAGS=$(CXXFLAGS)" "LDFLAGS=$(LDFLAGS)"'
+PKGDATAOPTS=-d .
 
 # Resource shortname
 PKGNAME=xercesc_messages_$(INTERFACE_VER_U)
