--- src/VBox/Main/Makefile.kmk.orig	2009-01-26 23:18:28.000000000 +0100
+++ src/VBox/Main/Makefile.kmk	2009-01-27 00:14:17.000000000 +0100
@@ -290,10 +290,11 @@
         win/HostPowerWin.cpp \
 	win/VBoxSVC.rc
 
+VBoxSVC_SOURCES.linux = \
+	linux/HostHardwareLinux.cpp
+
 ifdef VBOX_WITH_DBUS
-VBoxSVC_SOURCES.linux += \
-	linux/HostHardwareLinux.cpp \
-	linux/vbox-dbus.cpp
+  VBoxSVC_SOURCES.linux += linux/vbox-dbus.cpp
 endif
 
 VBoxSVC_SOURCES.solaris = \
